### DISCUSSION TITLE
请问有没有一个完整的微服务项目可供学习参考呢

### DISCUSSION THREAD

**Discussion by eric-gitta-moore** [sogou/workflow#1344] [Q&A] [answered]
请问有没有一个完整的微服务项目可供学习参考呢

希望能够是企业级一套解决方案

1、服务注册和发现
2、统一配置中心
3、路由网关
4、远程调用
5、熔断降级限流
6、分布式文件 Minio
7、统一权限控制
8、服务监控
9、链路跟踪 Skywalking 
10、消息中间件 RabbitMQ 
11、分布式任务
12、分布式事务
13、分布式日志 elk + kafka
14、支持 docker-compose、k8s、jenkins
15、CAS 单点登录 
16、路由限流

**Comment by holmes1412** [comment, marked as answer]
你好，感谢关注Workflow及生态项目，这是个非常好的主题。目前Workflow线上作为一套企业级的方案，有许多模块和生态项目可以分享，如下列出：

### 1、服务注册和发现
取决于用什么服务发现系统。以下是Workflow支持的几种：
1. 目前Workflow内部支持Consul客户端，可以直接对接Consul ： https://github.com/sogou/workflow/issues/1021
2. 支持腾讯的Polaris北极星（也是开源项目），需要使用workflow的生态项目接入：**[workflow-polaris](https://github.com/sogou/workflow-polaris)**
3. 支持接入Kubernetes的APIServer：**[workflow-k8s](https://github.com/sogou/workflow-k8s)**

### 2、统一配置中心
目前没有默认配置，因为各团队配置系统往往差别很大，而配置又是一个和模块交互相对较少的事项。目前Workflow的默认配置项很少，参考 **[about-config.md](https://github.com/sogou/workflow/blob/master/docs/about-config.md)** ，所以线上用法都是各团队用当前的配置系统把配置读出来之后设置到Workflow上接即可，这是最方便的解决办法。

### 3、路由网关
如果只需要基本的转发功能，可以参考：**[tutorial-05-http_proxy.md](https://github.com/sogou/workflow/blob/master/docs/tutorial-05-http_proxy.md)**，只要是Workflow或生态项目支持的协议，都可以做协议转换。不知道是否有其他需求呢？

###  4、远程调用
可以看看 **[SRPC](https://github.com/sogou/srpc)** ：目前也是 **[CNCF landscape](https://landscape.cncf.io/?selected=srpc)** 项目，是一个高性能轻量级RPC框架，支持protobuf、thrift，协议层面支持SRPC、BRPC、tRPC、Thirft协议及其对应的Http版本，也是截至目前唯一一个tRPC协议的开源版实现。

###  5、熔断降级限流
熔断等处理分两类：进城内和服务整体。前者在Workflow内自带，后者取决于你用什么服务发现与治理系统（参考第1点）。
Workflow内有一套完整的命名服务与治理体系：**[tutorial-15-name_service.md](https://github.com/sogou/workflow/blob/master/docs/tutorial-15-name_service.md)**
其中服务治理内包含了基本的熔断、半熔断和恢复功能：**[about-service-governance.md](https://github.com/sogou/workflow/blob/master/docs/about-service-governance.md)**

### 6、分布式文件 Minio
和上述多种外部系统一样，只要协议支持，Workflow就可以做基本的对接。Minio是S3协议，基于HTTP，你可以对WFHttpTask进行简单的封装就可以接入Minio了。

### 7、统一权限控制
Workflow目前要做权限控制可以在Server创建新链接时判断；生态项目SRPC有埋点，可以开发插件做鉴权。  
如果有整体权限控制，可能会涉及到接入鉴权系统，那么也取决于你的系统是什么协议，再看怎么接入。

### 8、服务监控
目前Workflow支持两种协议的监控上报：
1. Prometheus : 需要使用插件：**[wf-prometheus](https://github.com/sogou/wf-prometheus)**
2. OpenTelemetry : 需要使用SRPC框架的插件功能，参考：**[docs-09-metrics.md](https://github.com/sogou/srpc/blob/master/docs/docs-09-metrics.md)**

### 9、链路跟踪 Skywalking
目前Workflow使用SRPC里的插件，就可以上报到OpenTracing和Jaeger，也就是后来合并到OpenTelemetry的trace部分，参考：**[docs-08-tracing.md](https://github.com/sogou/srpc/blob/master/docs/docs-08-tracing.md)**

即使不使用RPC，只用Http，也可以用这些模块，需要参考这个文档：**[docs-10-http-with-modules.md](https://github.com/sogou/srpc/blob/master/docs/docs-10-http-with-modules.md)**

### 10、消息中间件 RabbitMQ 和 13、分布式日志 elk + kafka
分布式消息队列支持Kafka，接入方式参考：**[tutorial-13-kafka_cli.md](https://github.com/sogou/workflow/blob/master/docs/tutorial-13-kafka_cli.md)**，应该是最好用的C++异步Kafka客户端了。

### 11、分布式任务
上述Kafka例子内部实现就是一个分布式任务，你可以用Workflow实现任何分布式任务，当然Workflow是一个单进程内的概念，对于tensorflow这类分布式任务需要你针对自己的场景做封装。

### 12、分布式事务 和 15、CAS 单点登录
这两个不属于Workflow的功能内，可以参考上一条。

### 14、支持 docker-compose、k8s、jenkins
容器和框架本身是没有任何耦合关系的，Workflow和k8s的交互最多就是接入其API Server动态查找Pod，参考第一条。

### 16、路由限流
内部自带基本的路由功能，如果想用更灵活的RESTFul的API可以参考 **[wfrest](https://github.com/wfrest/wfrest)** 。

限流支持了**连接数限制**，可以通过配置修改。另外还有很多限制并发的做法，比如使用资源池，参考文档： **[about-resource-pool.md](https://github.com/sogou/workflow/blob/master/docs/about-resource-pool.md)** 。欢迎多翻翻issue：https://github.com/sogou/workflow/issues/1319

**Reply by eric-gitta-moore** [reply]
非常感谢您的详细回答，对我非常有指导性帮助，谢谢！
